partiql-logical-planner/src/typer.rs

```diff
@@ -0,0 +1,893 @@
+                    self.retrieve_type_from_local_ctx(&string_to_sym(OUTPUT_SCHEMA_KEY))
+                {
+                    self.output = Some(ty);
+                }
+            }
+            _ => self.errors.push(TypingError::NotYetImplemented(format!(
+                "Unsupported BindingOperator: {:?}",
+                &op
+            ))),
+        }
+    }
+
+    fn type_vexpr(&mut self, v: &ValueExpr, lookup_order: LookupOrder) {
+        fn to_bindings(s: &StructType) -> Vec<(SymbolPrimitive, PartiqlType)> {
+            s.fields()
+                .into_iter()
+                .map(|field| {
+                    let sym = SymbolPrimitive {
+                        value: field.name().to_string(),
+                        case: CaseSensitivity::CaseInsensitive,
+                    };
+
+                    (sym, field.ty().clone())
+                })
+                .collect()
+        }
+
+        fn binding_to_sym(binding: &BindingsName) -> SymbolPrimitive {
+            match binding {
+                BindingsName::CaseSensitive(s) => SymbolPrimitive {
+                    value: s.to_string(),
+                    case: CaseSensitivity::CaseSensitive,
+                },
+                BindingsName::CaseInsensitive(s) => SymbolPrimitive {
+                    value: s.to_string(),
+                    case: CaseSensitivity::CaseInsensitive,
+                },
+            }
+        }
+
+        match v {
+            ValueExpr::VarRef(binding_name, _) => {
+                let sym = binding_to_sym(binding_name);
+
+                match lookup_order {
+                    LookupOrder::GlobalLocal => {
+                        let ty = self.resolve_global_then_local(&sym);
+                        let ty = self.element_type(&ty);
+                        if ty.is_undefined() {
+                            self.type_with_undefined(&sym);
+                        } else {
+                            let mut new_type_env = LocalTypeEnv::new();
+                            if let TypeKind::Struct(s) = ty.kind() {
+                                to_bindings(s).into_iter().for_each(|b| {
+                                    new_type_env.insert(b.0, b.1);
+                                });
+
+                                let type_ctx = ty_ctx![(&new_type_env, &ty)];
+                                self.type_env_stack.push(type_ctx);
+                            } else {
+                                new_type_env.insert(sym, self.element_type(&ty));
+                                let type_ctx = ty_ctx![(&new_type_env, &ty)];
+                                self.type_env_stack.push(type_ctx);
+                            }
+                        }
+                    }
+                    LookupOrder::LocalGlobal => {
+                        let ty = self.resolve_local_then_global(&sym);
+
+                        if ty.is_undefined() {
+                            self.type_with_undefined(&sym);
```

Comment:
The difference is that the `resolve` methods return the types whereas `type_with_undefined` pushes the type in case of undefined to the stack. However I realized there is a simplification that can be made. Which is removing the `TypingMode` in the resolve methods. In the previous versions I was emitting an error in this code path and that one was removed but these where remaining. Thanks for pointing this out!